Protect the ResourceManager from division by zero when asked for 0 memory or 0 vcore

gray-matter opened the pull request:
When only one is 0, use the other value. If both are 0, return 0, which will indicate
that we don't want to allocate anything on that node.

```
java 2021-11-18 00:48:08,146 FATAL org.apache.hadoop.yarn.server.resourcemanager.ResourceManager: Error in handling event type NODE_UPDATE to the scheduler
java.lang.ArithmeticException: / by zero
	at org.apache.hadoop.yarn.util.resource.DominantResourceCalculator.computeAvailableContainers(DominantResourceCalculator.java:116)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.LeafQueue.assignContainer(LeafQueue.java:1542)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.LeafQueue.assignOffSwitchContainers(LeafQueue.java:1398)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.LeafQueue.assignContainersOnNode(LeafQueue.java:1277)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.LeafQueue.assignContainersInternal(LeafQueue.java:836)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.LeafQueue.assignContainers(LeafQueue.java:740)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainersToChildQueues(ParentQueue.java:590)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainers(ParentQueue.java:451)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainersToChildQueues(ParentQueue.java:590)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainers(ParentQueue.java:451)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainersToChildQueues(ParentQueue.java:590)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainers(ParentQueue.java:451)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.CapacityScheduler.allocateContainersToNode(CapacityScheduler.java:1113)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.CapacityScheduler.handle(CapacityScheduler.java:1168)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.CapacityScheduler.handle(CapacityScheduler.java:128)
	at org.apache.hadoop.yarn.server.resourcemanager.ResourceManager$SchedulerEventDispatcher$EventProcessor.run(ResourceManager.java:706)
	at java.lang.Thread.run(Thread.java:748)
```

gray-matter committed:
Protect the ResourceManager from division by zero when asked for 0 memory or 0 vcore

When only one is 0, use the other value. If both are 0, return 0, which will indicate
that we don't want to allocate anything on that node.

java 2021-11-18 00:48:08,146 FATAL org.apache.hadoop.yarn.server.resourcemanager.ResourceManager: Error in handling event type NODE_UPDATE to the scheduler
java.lang.ArithmeticException: / by zero
	at org.apache.hadoop.yarn.util.resource.DominantResourceCalculator.computeAvailableContainers(DominantResourceCalculator.java:116)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.LeafQueue.assignContainer(LeafQueue.java:1542)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.LeafQueue.assignOffSwitchContainers(LeafQueue.java:1398)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.LeafQueue.assignContainersOnNode(LeafQueue.java:1277)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.LeafQueue.assignContainersInternal(LeafQueue.java:836)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.LeafQueue.assignContainers(LeafQueue.java:740)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainersToChildQueues(ParentQueue.java:590)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainers(ParentQueue.java:451)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainersToChildQueues(ParentQueue.java:590)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainers(ParentQueue.java:451)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainersToChildQueues(ParentQueue.java:590)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.ParentQueue.assignContainers(ParentQueue.java:451)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.CapacityScheduler.allocateContainersToNode(CapacityScheduler.java:1113)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.CapacityScheduler.handle(CapacityScheduler.java:1168)
	at org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.CapacityScheduler.handle(CapacityScheduler.java:128)
	at org.apache.hadoop.yarn.server.resourcemanager.ResourceManager$SchedulerEventDispatcher$EventProcessor.run(ResourceManager.java:706)
	at java.lang.Thread.run(Thread.java:748)

ashangit commented:
FI on below commit which introduce a check on resources != 0 they are returning  Long.MAX_VALUE if both value are equal to 0 so probably better to do the same on your patch
```
Author: Wangda Tan <wangda@apache.org>
Date:   Fri Jan 29 10:53:31 2016 +0800

    YARN-4172. Extend DominantResourceCalculator to account for all resources. (Varun Vasudev via wangda)
       public long computeAvailableContainers(Resource available, Resource required) {
-    return Math.min(
-        available.getMemorySize() / required.getMemorySize(),
-        available.getVirtualCores() / required.getVirtualCores());
+    long min = Long.MAX_VALUE;
+    for (String resource : resourceNames) {
+      try {
+        ResourceInformation availableResource =
+            available.getResourceInformation(resource);
+        ResourceInformation requiredResource =
+            required.getResourceInformation(resource);
+        Long requiredResourceValue = UnitsConversionUtil
+            .convert(requiredResource.getUnits(), availableResource.getUnits(),
+                requiredResource.getValue());
+        Long tmp = availableResource.getValue() / requiredResourceValue;
+        min = min < tmp ? min : tmp;
+      } catch (YarnException ye) {
+        throw new IllegalArgumentException(
+            "Error getting resource information for " + resource, ye);
+      }
+
+    }
+    return min > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) min;
   }

```

Willymontaz commented:
@ashangit I don't agree with your reading of the code:

- requiredResourceValue will be 0 if the value is 0 (tested locally)
- thus tmp = availableResource.getValue() / requiredResourceValue; will fail with java.lang.ArithmeticException
- catch (YarnException ye) does not catch this exception, so the code fails with the same ArithmeticException

@gray-matter What I fear with your patch is that we allow the 0 value to continue its path down the code and possibly fail somewhere else.

ashangit commented:
@Willymontaz I indeed didn't push all the path of commit change but there is then another commit that introduce the 0 check
The one I put added the default min value to Long.MAX_VALUE
```
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/util/resource/DominantResourceCalculator.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/util/resource/DominantResourceCalculator.java
index 3c4413c4e75..7db1da44530 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/util/resource/DominantResourceCalculator.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/util/resource/DominantResourceCalculator.java
@@ -183,8 +183,10 @@ public long computeAvailableContainers(Resource available, Resource required) {
         Long requiredResourceValue = UnitsConversionUtil
             .convert(requiredResource.getUnits(), availableResource.getUnits(),
                 requiredResource.getValue());
-        Long tmp = availableResource.getValue() / requiredResourceValue;
-        min = min < tmp ? min : tmp;
+        if (requiredResourceValue != 0) {
+          Long tmp = availableResource.getValue() / requiredResourceValue;
+          min = min < tmp ? min : tmp;
+        }
       } catch (YarnException ye) {
         throw new IllegalArgumentException(
****
```

My concern with computeAvailableContainers returning 0 is the possibility to have containers never allocated meaning kept in memory in pending state by the RM (and also stored in the zookeeper for HA)
But I can be wrong

Willymontaz commented:
Ok, indeed the final code in HDP3 looks like this  (which implies value returned is Long.MAX_VALUE as you stated

long min = Long.MAX_VALUE;
    int maxLength = ResourceUtils.getNumberOfCountableResourceTypes();
    for (int i = 0; i < maxLength; i++) {
      ResourceInformation availableResource = available
          .getResourceInformation(i);
      ResourceInformation requiredResource = required.getResourceInformation(i);
      if (requiredResource.getValue() != 0) {
        long tmp = availableResource.getValue() / requiredResource.getValue();
        min = min < tmp ? min : tmp;
      }
    }
    return min > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) min;

gray-matter committed:
Add a warning when encountering a reservation with 0 memory 0 vcore

gray-matter committed:
Return Integer.MAX_VALUE when computeAvailableContainers is required 0 vcore 0 memory

This is what other patches that try to fix this issue do.
It feels a bit counterintuitive reading the code, but they probably have
their reasons and this code will probably never run anyway.

Willymontaz commented on hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/util/resource/DominantResourceCalculator.java:
```diff
@@ -118,8 +121,10 @@ public int computeAvailableContainers(Resource available, Resource required) {
     if (required.getMemory() == 0) {
-      if (required.getVirtualCores() == 0)
+      if (required.getVirtualCores() == 0) {
+        LOG.warn("Got a resources reservation with 0 vcore and 0 memory, acting as if no resource was available");
```
We will only get a warn and not much context. Maybe we could verify memory and vcores before entering computeAvailableContainers / or use exception based mechanism to log more info about the request that led to this -> application_id and maybe the complete request.

gray-matter replied:
Agree, that's a good idea since there is actually only one relevant path

gray-matter committed:
Add more context for the 0 vcore 0 memory request

Targeting the path that we saw in the previous error reports
(there are 2 callers for this method)
